subs/pantry/src/Pantry/Storage.hs

```diff
@@ -436,30 +449,121 @@ loadHackageTarballInfo name version = do
   where
     go (Entity _ ht) = (hackageTarballSha ht, hackageTarballSize ht)
 
+getKey :: Either (Entity record) (Key record) -> Key record
+getKey dat = case dat of
+               Left (Entity key _) -> key
+               Right key -> key
+
+storeCabalFile ::
+       (HasPantryConfig env, HasLogFunc env, HasProcessContext env)
+    => ByteString
+    -> P.PackageName
+    -> ReaderT SqlBackend (RIO env) BlobId
+storeCabalFile cabalBS pkgName = do
+    (bid, _) <- storeBlob cabalBS
+    let cabalFile = P.cabalFileName pkgName
+    _ <- insertBy FilePath {filePathPath = cabalFile}
+    return bid
+
+loadFilePath ::
+       (HasPantryConfig env, HasLogFunc env, HasProcessContext env)
+    => SafeFilePath
+    -> ReaderT SqlBackend (RIO env) (Entity FilePath)
+loadFilePath path = do
+    fp <- getBy $ UniqueSfp path
+    case fp of
+        Nothing ->
+            error $
+            "loadFilePath: No row found for " <>
+            (T.unpack $ P.unSafeFilePath path)
+        Just record -> return record
+
+generateHPack ::
+       (HasPantryConfig env, HasLogFunc env, HasProcessContext env)
+    => TreeId
+    -> VersionId
+    -> ReaderT SqlBackend (RIO env) (Key HPack)
+generateHPack tid vid = do
+    let hpackPath =
+            fromMaybe
+                (error
+                     "generateHPack: Not able to convert hpack file to SafeFilePath")
+                (P.mkSafeFilePath (T.pack Hpack.packageConfig))
+    filepath <- loadFilePath hpackPath
+    let filePathId :: FilePathId = entityKey filepath
+    hpackTreeEntry <-
+        selectFirst [TreeEntryTree ==. tid, TreeEntryPath ==. filePathId] []
+    hpackEntity <-
+        case hpackTreeEntry of
+            Nothing ->
+                error $
+                "generateHPack: No package.yaml file found in TreeEntry for TreeId:  " ++
+                (show tid)
+            Just record -> return record
+    hpackBS <- loadBlobById (treeEntryBlob $ entityVal hpackEntity)
+    (pkgName, cabalBS) <- lift $ hpackToCabal hpackBS
+    bid <- storeCabalFile cabalBS pkgName
+    let cabalFile = P.cabalFileName pkgName
+    fid <- insertBy FilePath {filePathPath = cabalFile}
+    let hpackRecord =
+            HPack
+                { hPackTree = tid
+                , hPackVersion = vid
+                , hPackCabal = bid
+                , hPackPath = getKey fid
+                }
+    val <- insertBy hpackRecord
+    return $ getKey val
+
+
+hpackVersionId ::
+       (HasPantryConfig env, HasLogFunc env, HasProcessContext env)
+    => ReaderT SqlBackend (RIO env) VersionId
+hpackVersionId = do
+    hpackSoftwareVersion <- lift $ hpackVersion
```

Comment:
Does this handle the case when hpack is an external program? Doesn't this assume we're always using the compiled-in version?

Reply:
Yes, it does handle the case when `hpack` is an external program in the implementation of the function `hpackVersion`. We pattern match between `HpackBundled` and `HpackCommand` for it.

Reply:
Got it, thanks for clarifying.